### PR TITLE
Refactor GetGrantCaptcha to use callback

### PR DESCRIPTION
**Pull request by jdkuki**
Resolves brave/brave-browser#5140

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 1]
```diff
@@ -555,24 +555,19 @@ - (void)onGrant:(ledger::Result)result grant:(ledger::GrantPtr)grant
...
+  ledger->GetGrantCaptcha(headers,
+      ^(const std::string &image, const std::string &hint) {
+        completion([NSString stringWithUTF8String:image.c_str()],
+                   [NSString stringWithUTF8String:hint.c_str()]);
+      });
 }
 
 - (void)onGrantCaptcha:(const std::string &)image hint:(const std::string &)hint
 {
-  if (self.grantCaptchaBlock) {
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: You can delete the `grantCaptchaBlock` @property definition now that its no longer used. 

Also can you just add a comment here saying this is unused, better than fully empty I guess 😛

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 2]
```diff
@@ -555,24 +555,19 @@ - (void)onGrant:(ledger::Result)result grant:(ledger::GrantPtr)grant
...
-               [NSString stringWithUTF8String:hint.c_str()]);
-  };
   std::vector<std::string> headers;
   headers.push_back("brave-product:brave-core");
   headers.push_back("promotion-id:" + std::string(promoID.UTF8String));
   headers.push_back("promotion-type:" + std::string(promotionType.UTF8String));
-  ledger->GetGrantCaptcha(headers);
+  ledger->GetGrantCaptcha(headers,
+      ^(const std::string &image, const std::string &hint) {
+        completion([NSString stringWithUTF8String:image.c_str()],
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: Is grant captcha doing an asynchronous network call? I can't recall, but if it is, can we wrap this completion block in a `dispatch_async` to main thread as seen in other callbacks

File: vendor/brave-ios/Ledger/BATBraveLedger.mm (hunk 1 above)
Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: Fixed

File: vendor/brave-ios/Ledger/BATBraveLedger.mm (hunk 2 above)
Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: Fixed

File: components/services/bat_ledger/bat_ledger_impl.cc [hunk 3]
```diff
@@ -184,9 +184,25 @@ void BatLedgerImpl::FetchGrants(const std::string& lang,
   ledger_->FetchGrants(lang, payment_id);
 }
 
+
+// static
+void BatLedgerImpl::OnGetGrantCaptcha(
+    CallbackHolder<GetGrantCaptchaCallback>* holder,
+    std::string image,
+    std::string hint) {
+  if (holder->is_valid())
```
Comment (emerick) [inline components/services/bat_ledger/bat_ledger_impl.cc]: Let's confirm that holder isn't null before accessing it here.

File: components/services/bat_ledger/public/interfaces/bat_ledger.mojom [hunk 4]
```diff
@@ -63,7 +63,7 @@ interface BatLedger {
       int32 month, int32 year, uint32 data);
 
   FetchGrants(string lang, string payment_id);
-  GetGrantCaptcha(array<string> headers);
+  GetGrantCaptcha(array<string> headers) =>(string image, string hint);
```
Comment (emerick) [inline components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: Nit: Should have a space after the arrow.

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h [hunk 5]
```diff
@@ -50,6 +50,8 @@ using GetAddressesCallback =
 using HasSufficientBalanceToReconcileCallback = std::function<void(bool)>;
 using FetchBalanceCallback = std::function<void(ledger::Result,
                                                 ledger::BalancePtr)>;
+using GetGrantCaptchaCallback = std::function<void(std::string,
```
Comment (emerick) [inline vendor/bat-native-ledger/include/bat/ledger/ledger.h]: Same comment as above: let's use `const std::string&` for these parameters if possible.

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 6]
```diff
@@ -733,8 +733,9 @@ void LedgerImpl::OnGrant(ledger::Result result,
...
 
 void LedgerImpl::GetGrantCaptcha(
-    const std::vector<std::string>& headers) const {
-  bat_grants_->GetGrantCaptcha(headers);
+    const std::vector<std::string>& headers,
+    ledger::GetGrantCaptchaCallback callback) const {
+  bat_grants_->GetGrantCaptcha(headers, std::move(callback));
 }
 
 void LedgerImpl::OnGrantCaptcha(const std::string& image,
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: please remove it as it's not needed anymore

File: components/services/bat_ledger/bat_ledger_impl.cc [hunk 7]
```diff
@@ -184,9 +184,25 @@ void BatLedgerImpl::FetchGrants(const std::string& lang,
   ledger_->FetchGrants(lang, payment_id);
 }
 
+
+// static
+void BatLedgerImpl::OnGetGrantCaptcha(
+    CallbackHolder<GetGrantCaptchaCallback>* holder,
+    std::string image,
```
[the remark this answers is not in the document]
Comment (jdkuki) [reply components/services/bat_ledger/bat_ledger_impl.cc]: fixed

File: components/services/bat_ledger/bat_ledger_impl.cc (hunk 3 above)
Comment (jdkuki) [reply components/services/bat_ledger/bat_ledger_impl.cc]: added `DCHECK(holder)`

File: components/services/bat_ledger/public/interfaces/bat_ledger.mojom (hunk 4 above)
Comment (jdkuki) [reply components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: fixed

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h (hunk 5 above)
Comment (jdkuki) [reply vendor/bat-native-ledger/include/bat/ledger/ledger.h]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 6 above)
Comment (jdkuki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 8]
```diff
@@ -737,13 +737,9 @@ void LedgerImpl::OnGrant(ledger::Result result,
 }
 
 void LedgerImpl::GetGrantCaptcha(
-    const std::vector<std::string>& headers) const {
-  bat_grants_->GetGrantCaptcha(headers);
-}
-
-void LedgerImpl::OnGrantCaptcha(const std::string& image,
-                                const std::string& hint) {
-  ledger_client_->OnGrantCaptcha(image, hint);
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: I don't see this removed from `ledger_client.h` and then all other places. Do we use it somewhere else?

**Comment by jdkuki** [comment]
> Ledger_client.h check

`OnGrantCaptcha` is the callback I am passing down from `rewards_service` to `ledger`

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 8 above)
Comment (jdkuki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: OnGrantCaptcha is the callback I am passing down from rewards_service to ledger

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 9]
```diff
@@ -1402,7 +1402,8 @@ void RewardsServiceImpl::GetGrantCaptcha(
   headers.push_back("brave-product:brave-core");
   headers.push_back("promotion-id:" + promotion_id);
   headers.push_back("promotion-type:" + promotion_type);
-  bat_ledger_->GetGrantCaptcha(headers);
+  bat_ledger_->GetGrantCaptcha(headers,
+      base::BindOnce(&RewardsServiceImpl::OnGrantCaptcha, AsWeakPtr()));
```
Comment (jdkuki) [inline components/brave_rewards/browser/rewards_service_impl.cc]: OnGrantCaptcha used here

File: vendor/brave-ios/Ledger/Generated/NativeLedgerClient.mm [hunk 10]
```diff
@@ -109,9 +106,7 @@
 }
 void NativeLedgerClient::SaveActivityInfo(ledger::PublisherInfoPtr publisher_info, ledger::PublisherInfoCallback callback) {
   [bridge_ saveActivityInfo:std::move(publisher_info) callback:callback];
-}
-void NativeLedgerClient::SaveContributionInfo(const std::string & probi, const int month, const int year, const uint32_t date, const std::string & publisher_key, const ledger::REWARDS_CATEGORY category) {
-  [bridge_ saveContributionInfo:probi month:month year:year date:date publisherKey:publisher_key category:category];
+} void NativeLedgerClient::SaveContributionInfo(const std::string & probi, const int month, const int year, const uint32_t date, const std::string & publisher_key, const ledger::REWARDS_CATEGORY category) { [bridge_ saveContributionInfo:probi month:month year:year date:date publisherKey:publisher_key category:category];
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/Generated/NativeLedgerClient.mm]: This seems to have lost a new line, but I'll fix it later

**Pull request merged**